drivers/input/input_pinnacle.c

```diff
@@ -0,0 +1,627 @@
+	sample->x = ((rx[2] & 0x0F) << 8) | rx[0];
+	sample->y = ((rx[2] & 0xF0) << 4) | rx[1];
+	sample->z = rx[3] & 0x3F;
+#elif defined(CONFIG_INPUT_PINNACLE_DATA_MODE_RELATIVE)
+	sample->x = ((rx[0] & BIT(4)) == BIT(4)) ? -(256 - rx[1]) : rx[1];
+	sample->y = ((rx[0] & BIT(5)) == BIT(5)) ? -(256 - rx[2]) : rx[2];
+#endif
+}
+
+#if defined(CONFIG_INPUT_PINNACLE_DATA_MODE_ABSOLUTE)
+static bool pinnacle_is_idle_sample(const struct pinnacle_sample *sample)
+{
+	return (sample->x == 0 && sample->y == 0 && sample->z == 0);
+}
+
+static void pinnacle_clip_sample(const struct pinnacle_active_range *range,
+				 struct pinnacle_sample *sample)
+{
+	if (sample->x < range->x_min) {
+		sample->x = range->x_min;
+	}
+	if (sample->x > range->x_max) {
+		sample->x = range->x_max;
+	}
+	if (sample->y < range->y_min) {
+		sample->y = range->y_min;
+	}
+	if (sample->y > range->y_max) {
+		sample->y = range->y_max;
+	}
+}
+
+static void pinnacle_scale_sample(const struct pinnacle_active_range *range,
+				  const struct pinnacle_resolution *res,
+				  struct pinnacle_sample *sample)
+{
+	sample->x = (uint16_t)((uint32_t)(sample->x - range->x_min) * res->x /
+			       (range->x_max - range->x_min));
+	sample->y = (uint16_t)((uint32_t)(sample->y - range->y_min) * res->y /
+			       (range->y_max - range->y_min));
+}
+#endif
+
+static int pinnacle_sample_fetch(const struct device *dev, struct pinnacle_sample *sample)
+{
+	const struct pinnacle_config *config = dev->config;
+	const struct pinnacle_bus *bus = &config->bus;
+
+	int ret1 = 0;
+	int ret2 = 0;
+
+#if defined(CONFIG_INPUT_PINNACLE_DATA_MODE_ABSOLUTE)
+	uint8_t rx[4];
+
+	ret1 = bus->seq_read(bus, PINNACLE_REG_PACKET_BYTE2, rx, 4);
+#elif defined(CONFIG_INPUT_PINNACLE_DATA_MODE_RELATIVE)
+	uint8_t rx[3];
+
+	ret1 = bus->seq_read(bus, PINNACLE_REG_PACKET_BYTE0, rx, 3);
+#endif
+
+	if (ret1) {
+		LOG_ERR("Failed to read data from SPI device");
+	} else {
+		pinnacle_decode_sample(rx, sample);
+	}
+	ret2 = bus->write(bus, PINNACLE_REG_STATUS1, 0x00);
+	if (ret2) {
+		LOG_ERR("Failed to clear SW_CC and SW_DR for SPI device");
+		return ret2;
+	}
+	return 0;
+}
+
+static int pinnacle_handle_interrupt(const struct device *dev)
+{
+	struct pinnacle_data *drv_data = dev->data;
+	struct pinnacle_sample *sample = &drv_data->sample;
+
+	int ret = pinnacle_sample_fetch(dev, sample);
+
+	if (ret) {
+		LOG_ERR("Failed to read data packets");
+	}
+
+#if defined(CONFIG_INPUT_PINNACLE_DATA_MODE_ABSOLUTE)
+	const struct pinnacle_config *config = dev->config;
+
+	if (config->clipping_enabled && !pinnacle_is_idle_sample(sample)) {
+		pinnacle_clip_sample(&config->active_range, sample);
+		if (config->scaling_enabled) {
+			pinnacle_scale_sample(&config->active_range, &config->resolution, sample);
+		}
+	}
+
+	input_report_abs(dev, INPUT_ABS_X, sample->x, false, K_FOREVER);
+	input_report_abs(dev, INPUT_ABS_Y, sample->y, false, K_FOREVER);
+	input_report_abs(dev, INPUT_ABS_Z, sample->z, true, K_FOREVER);
```

Comment:
Yeah I'm thinking the same, this could probably have a threshold and report a touch. It could be handled at application level as well though, up to you.

By the way, should it not be reported for relative mode as well? Like you'd report rel-x, rel-y and either touch or abs-x, like a normal touchpad.

Reply:
The controller supports two modes: absolute and relative. The mode can be configured in the `FeedConfig1` register by setting 0 for absolute and 1 for relative in the `Data mode` BIT(1). In the absolute level, it provides X and Y coordinates and Z level. In the relative mode, it provides X and Y coordinates without providing Z level. Also, in this mode there are certain features which can be enabled in the `FeedConfig2` register, such as scroll, primary and secondary taps and an extended area. The absolute mode is considered to be more flexible because the application can decide how to implement all these features on its own, where as the relative mode should give a trackpad with all features working from a box. I tried the relative mode, the native one in the controller, and found that it works strangely: it doesn't consistently detect finer motions. I have a model with a glued curved overlay, I could assume that by default the sensitivity is not tuned to properly work with this overlay. I'm mostly focusing on the absolute mode right now and adding the native relative mode in order to explore its capabilities. So, I think that there are two ways how to implement the relative mode: (1) the driver can configure the controller to report absolute coordinates and do all necessarily calculations to provide relative coordinates and taps, (2) the driver can configure the controller to report relative coordinates and simply report them. What do you think?

Reply:
Very interesting, that sent me down a rabbit hole looking for teardowns of the Steam Controller (that apparently is using this thing) and it turns out that they actually have a tactile switch after each pad, so i guess they may be using relative mode with the switch for the press event and that's probably the indended implementation.

I also have a curved overlay model, was playing with your driver and found it was very inconsistent, @petejohanson mentioned to me that he found the same and he had to do some sensitivity adjustment:

https://github.com/zmkfirmware/zmk/pull/2035/commits/6f060737847a36a0421015dd675de94b9f9dd6c3#diff-787362a61e899da2a4d0d797544fa5cc524751ee0d08d6ac77777b51e4c53086R317-R318

can we try that one? and keep the two modes, I like that the driver have the option now

Reply:
I implemented an ability to configure the sensor sensitivity and it makes a huge difference, you should try! I used the official documentation GT-AN-090623 to figure out how to work with extended registers and also an official example to figure out the right register, bits and possible values. The example says by default ASIC is configured for the lowest sensitivity to properly work with flag overlays. So, higher sensitivity is needed for curved ones, I found that 3x and 2x are much better than then default 4x for them. I kept the default value set for 4x, so please pay attention that the default is not good for curved overlays. I also found a bug in decoding relative coordinates, I'll work on fixing this in follow up patches.